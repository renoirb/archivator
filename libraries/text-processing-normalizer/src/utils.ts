/* eslint-disable @typescript-eslint/ban-ts-ignore */

import { RecordToMapFactoryType } from './types'

/**
 * Sorting closure.
 *
 * Assuming each value are some sort of count.
 * We want the biggest number first.
 *
 * When matching Array from following signature
 *
 * The following should match an array of type `Array<[string, number]>`
 *
 * Here is copy-pasta from VSCode for compare function.
 *
 * ```
 * Array<[string, number]>.sort(compareFn?: ((a: [string, number], b: [string, number]) => number) | undefined): [string, number][]
 * ```
 *
 * There must be a way to extract `compareFn` argument type using TypeScriptâ€™s utility types.
 * I haven't figured it out yet.
 *
 */
const whenValuesAreNumbersFromBiggestToLowest = (
  a: [string, number],
  b: [string, number],
): number => -1 * (a[1] - b[1])

export const sorting = {
  whenValuesAreNumbersFromBiggestToLowest,
}

export const convertRecordHashMapToMap: RecordToMapFactoryType<
  string,
  number
> = input => {
  const map = new Map<string, number>(Object.entries(input))

  const out = new Map<string, number>(
    [...map.entries()].sort(sorting.whenValuesAreNumbersFromBiggestToLowest),
  )

  return out
}
