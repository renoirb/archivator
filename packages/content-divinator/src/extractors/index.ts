export * from './words'
export * from './summary'
