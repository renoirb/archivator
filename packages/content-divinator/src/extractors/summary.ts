import { SummaryRecordType, WordUsageMapType } from '../types'

/**
 * Summarizing content.
 *
 * For starters, what are the most used "keywords".
 * More things to come.
 *
 * ----
 *
 * See earlier implementation:
 *   v2.0.0, rework implementation:
 *     links:
 *       - name: sortedAndKeywords
 *         replacing: analyze
 *         replacedWith: summary
 *         url: https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L50-L62
 *
 * ----
 *
 * @package extractors
 *
 * @param wordsMap - Unique "word" where each value is its usage count
 * @param floor - In keywords grouping, what is the minimum number of times to qualify (default: 3)
 * @param max - In keywords grouping, how many items in the top keywords (default: 10)
 */
export const summary = (
  wordsMap: WordUsageMapType,
  floor: number = 3,
  max: number = 10,
): SummaryRecordType => {
  const zeroIndexSafeFloor = Number.isInteger(+floor) && +floor > 0 ? floor : 1
  const keywords: string[] = []
  let iter = 0
  // @ts-ignore
  for (const [word, count] of wordsMap.entries()) {
    if (iter < max && count > zeroIndexSafeFloor) {
      keywords.push(word)
    }
    iter++
  }

  const out: SummaryRecordType = { keywords }

  return out
}
