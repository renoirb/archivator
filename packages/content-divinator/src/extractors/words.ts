import {
  wordNormalizer,
  nonStopWordIsser,
  convertRecordHashMapToMap,
} from '../utils'
import type { WordUsageMapType } from '../types'

/**
 * Extract words and calculate usage frequency.
 *
 * @internal
 * @package extractors
 *
 * {@link https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L22-L46 replacing v1.0 implementation of extractWords}
 * {@link https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L76-L94 replacing v1.0 implementation of analyze}
 * {@link https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L63-L74 replacing v1.0 implementation of sort}
 * {@link https://github.com/renoirb/archivator/blob/v1.0.0/src/analyze.js#L18-L20 replacing v1.0 implementation of removePunctuation}
 * {@link https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L13-L16 replacing v2.0 implementation of normalize as wordNormalizer}
 * {@link https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L18-L48 replacing v2.0 implementation of extractWords as words}
 * {@link https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L64-L77 replacing v2.0 implementation of sort}
 *
 * @param body - Text content, as a single string
 * @param stopWords - Words that should be ignored
 * @param locales - Locales tags to support for {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLocaleLowerCase|String.prototype.toLocaleLowerCase} (default: `['en-ca']`)
 */
export const words = (
  body: string,
  stopWords: string[] = [],
  locales: string | string[] = ['en-CA'],
): WordUsageMapType => {
  const normalizeWord = wordNormalizer(locales)
  const isNonStopWord = nonStopWordIsser(stopWords)
  // We want to use Object.create(null) instead of Map
  // because we want to quickly increment values on an Hash-Map
  const words: Record<string, number> = Object.create(null)
  const foundOnce = new Set<string>()
  const text = body.split(' ')
  for (let i = 0; i < text.length; i++) {
    const word = normalizeWord(text[i])
    if (word && isNonStopWord(word)) {
      if (foundOnce.has(word)) {
        // This was the former way to check, before we had Map.
        // Leaving here for old memoriesâ€™ sake.
        if (Object.prototype.hasOwnProperty.call(words, word)) {
          words[word]++
        } else {
          words[word] = 2
        }
      } else {
        foundOnce.add(word)
      }
    }
  }

  const sorted = convertRecordHashMapToMap(words)

  return Object.seal(sorted)
}
