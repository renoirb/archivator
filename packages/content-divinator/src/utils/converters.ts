import { sorting } from '../sorting'
import { RecordToMapFactoryType, MapToRecordHashMapFactoryType } from '../types'

/**
 * Convert `Record<string, number>` HashMap to `Map<string, number>`
 *
 * For sorting, refer to {@link sorting.whenRecordValueIsNumberOrderByDescending}
 *
 * @package utils
 *
 * @param textHashMap - Unique "word" where each value is its usage count
 */
export const convertRecordHashMapToMap: RecordToMapFactoryType<
  string,
  number
> = textHashMap => {
  const map = new Map<string, number>(Object.entries(textHashMap))

  const out = new Map<string, number>(
    [...map.entries()].sort(sorting.whenRecordValueIsNumberOrderByDescending),
  )

  return out
}

/**
 * Convert `Map<string, number>` to `Record<string, number>` HashMap
 *
 * @public
 * @package utils
 *
 * @param textHashMap - Unique "word" where each value is its usage count
 */
export const convertMapToRecordHashMap: MapToRecordHashMapFactoryType<
  string,
  number
> = textMap => {
  const textHashMap: Record<string, number> = Object.create(null)

  for (let [key, value] of textMap.entries()) {
    Object.assign(textHashMap, { [key]: value })
  }

  return textHashMap
}
