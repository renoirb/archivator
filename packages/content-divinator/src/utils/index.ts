export * from './converters'
export * from './predicates'
