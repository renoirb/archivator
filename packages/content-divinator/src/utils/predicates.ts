import { WordNormalizerType, NonStopWordIsserType } from '../types'

/**
 * Remove any non alpha numeric items.
 *
 * @package utils
 *
 * ----
 *
 * Bookmarks:
 * - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLocaleLowerCase
 *
 * See:
 * - https://github.com/renoirb/archivator/blob/v2.0.0/src/analyze.js#L13-L16
 */
export const wordNormalizer = (
  locales?: string | string[],
): WordNormalizerType => word =>
  word && typeof word === 'string'
    ? word.replace(/[^\w]/g, '').toLocaleLowerCase(locales)
    : void 0

/**
 * Utility method to shave off words with no meaning.
 *
 * @public
 * @package utils
 */
export const nonStopWordIsser = (
  stopWords: string[] = [],
): NonStopWordIsserType => {
  /**
   * https://www.ranks.nl/stopwords
   * http://xpo6.com/list-of-english-stop-words/
   */
  const stopWordsSet = new Set([...stopWords])

  return word => stopWordsSet.has(word) === false
}
