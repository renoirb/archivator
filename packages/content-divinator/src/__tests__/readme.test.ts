/* eslint-disable @typescript-eslint/ban-ts-ignore */

import { ContentDivinator } from '..'
import { summary } from '../extractors'
import { convertMapToRecordHashMap } from '../utils'

describe('README.md examples', () => {
  test('ContentDivinator', () => {
    /**
     * Let's attempt to wordsMap that looks like this:
     *
     * The following is a `Record<string, number>` representation of `Map<string, number>`.
     *
     * Refer to {@link convertMapToRecordHashMap|convertMapToRecordHashMap} for conversion algorithm.
     *
     * So we could be able to guess what's the main topic.
     */
    const desiredTextHashMap = {
      chuck: 5,
      wood: 4,
      woodchuck: 4,
    }
    const textContent = `
      How much wood would a woodchuck chuck
      if a woodchuck could chuck wood?
      He would chuck, he would, as much as he could,
      and chuck as much wood as a woodchuck would
      if a woodchuck could chuck wood.
    `
    // Letâ€™s remove a few words for simplicity.
    const stopWords = ['a', 'as', 'could', 'he', 'how', 'if', 'much', 'would']
    const subject = new ContentDivinator(stopWords)
    const wordsMap = subject.words(textContent)
    expect(wordsMap).toMatchSnapshot()
    const summarized = summary(wordsMap, 3)
    expect(summarized).toMatchSnapshot()
    expect(wordsMap.get('chuck')).toBe(5)
    expect(wordsMap.get('wood')).toBe(4)
    expect(wordsMap.get('woodchuck')).toBe(4)
    const textHashMap = convertMapToRecordHashMap(wordsMap)
    expect(textHashMap).toMatchSnapshot()
    expect(textHashMap).toMatchObject(desiredTextHashMap)
  })
})
