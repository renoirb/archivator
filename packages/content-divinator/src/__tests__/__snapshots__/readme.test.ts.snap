// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`README.md examples ContentDivinator 1`] = `
Map {
  "chuck" => 5,
  "woodchuck" => 4,
  "wood" => 4,
}
`;

exports[`README.md examples ContentDivinator 2`] = `
Object {
  "keywords": Array [
    "chuck",
    "woodchuck",
    "wood",
  ],
}
`;

exports[`README.md examples ContentDivinator 3`] = `
Object {
  "chuck": 5,
  "wood": 4,
  "woodchuck": 4,
}
`;
